### PR TITLE
Backport months of work to 4.1

### PR DESCRIPTION
**Pull request by jaredbeck**
@batter 

This took a little while to prepare, and involved a few conflicts. In particular, I had to take care to avoid two things:

1. Anything incompatible with ruby 1.8, which we still support in PT 4.1
2. The other breaking change in PT 5.0, namely the removal of the `set_paper_trail_whodunnit` callback. https://github.com/airblade/paper_trail/pull/556

Please let me know if I included anything I shouldn't have.

### CODE REVIEW

**Comment by jaredbeck** [comment]
Hmm .. I'll have to pull the rubocop stuff out of here, as rubocop requires ruby >= 1.9.3.  I'll do that later tonight.

**Comment by jaredbeck** [comment]
Whew! I had to fix a few things that were incompatible with ruby 1.8.7, but I think this is finally ready for review.

**Comment by jaredbeck** [comment]
Thanks to the eight humans and one robot (@orthographic-pedant) who have contributed so far to PT 4.1! :tada: :balloon:

**Comment by batter** [comment]
Hey @jaredbeck, I'm on holiday with my family this week so haven't had a ton of time to look at this stuff but will try to take a peek tonight. Looks good on the surface though, I'm guessing you would know better than I what makes sense to backport.

**Comment by jaredbeck** [comment]
> Looks good on the surface though, I'm guessing you would know better than I what makes sense to backport.

Thanks, just let me know if you see anything that looks like a breaking change.

Enjoy the holiday! :christmas_tree:

**Comment by jaredbeck** [comment]
Mind if I merge this, Ben?

File: lib/paper_trail.rb [hunk 1]
```diff
@@ -126,9 +128,9 @@ def self.paper_trail_store
 
   # Returns PaperTrail's configuration object.
   def self.config
-    @@config ||= PaperTrail::Config.instance
-    yield @@config if block_given?
-    @@config
+    @config ||= PaperTrail::Config.instance
```
Comment (batter) [inline lib/paper_trail.rb]: Was there a pro / con to storing this as an instance or class level variable? Is this being done to address thread safety concerns?

Comment (jaredbeck) [inline lib/paper_trail.rb]: It was changed from a **class variable** to a **class instance variable**. The change is purely stylistic, and has no effect on thread safety. See https://github.com/airblade/paper_trail/pull/641 for rationale.

**Comment by jaredbeck** [comment]
@kaspth Kasper, you had said you were using a fork of `4.0-stable` plus https://github.com/airblade/paper_trail/pull/667 in production. Can you please try out `4.1-stable` now that https://github.com/airblade/paper_trail/pull/667 has been backported? Thanks. I think we're almost ready for a 4.1.0 release.

**Comment by batter** [comment]
Sorry @jaredbeck, never had a chance to go through all of the changes but everything looked like it was on point from what I did look through. Nice work. Can try to do some testing with you at some point next week if helpful.

**Comment by jaredbeck** [comment]
No worries, Ben. Hopefully a few people can try out the `4.1-stable` branch before we cut a 4.1.0 release. I have one app at work I can test with, but it doesn't track associations.

**Comment by kaspth** [comment]
@jaredbeck sweet! I'll try pointing our Gemfile at `4.1-stable` next week :grin:

**Pull request merged**